nixos/modules/services/backup/tsm.nix

```diff
@@ -0,0 +1,301 @@
+      The password will be stored in the directory
+      given by the option <option>passwdDir</option>.
+      <emphasis>Caution</emphasis>:
+      If this option is enabled and the server forces
+      to renew the password (e.g. on first connection),
+      a random password will be generated and stored
+    '';
+    options.passwdDir = lib.options.mkOption {
+      type = lib.types.path;
+      example = "/home/alice/tsm-password";
+      description = ''
+        Directory that holds the TSM
+        node's password information.
+        The value will be used for the
+        <literal>passworddir</literal>
+        directive in <filename>dsm.sys</filename>.
+      '';
+    };
+    options.includeExclude = lib.options.mkOption {
+      type = lib.types.lines;
+      default = "";
+      example = ''
+        exclude.dir     /nix/store
+        include.encrypt /home/.../*
+      '';
+      description = ''
+        <literal>include.*</literal> and
+        <literal>exclude.*</literal> directives to be
+        used when sending files to the IBM TSM server.
+        The lines will be written into a file that the
+        <literal>inclexcl</literal>
+        directive in <filename>dsm.sys</filename> points to.
+
+      '';
+    };
+    options.extraConfig = lib.options.mkOption {
+      # TSM option keys are case insensitive;
+      # we have to ensure there are no keys that
+      # differ only by upper and lower case.
+      type = lib.types.addCheck
+        (lib.types.attrsOf (lib.types.nullOr lib.types.str))
+        (attrs: checkIUnique (lib.attrsets.attrNames attrs));
+      default = {};
+      example.compression = "yes";
+      description = ''
+        Additional key-value pairs for the server stanza.
+        Values must be strings, or <literal>null</literal>
+        for the key not to be used in the stanza
+        (e.g. to overrule values generated by other options).
+      '';
+    };
+    options.stanza = lib.options.mkOption {
+      type = lib.types.str;
+      internal = true;
+      visible = false;
+      description = "Server stanza text generated from the options.";
+    };
+    config.name = lib.modules.mkDefault name;
+    # Client system-options file directives are explained here:
+    # https://www.ibm.com/support/knowledgecenter/SSEQVQ_8.1.7/client/c_opt_usingopts.html
+    config.extraConfig =
+      lib.attrsets.mapAttrs (lib.trivial.const lib.modules.mkDefault) (
+        {
+          commmethod = "v6tcpip";  # uses v4 or v6, based on dns lookup result
+          tcpserveraddress = config.server;
+          tcpport = builtins.toString config.port;
+          nodename = config.node;
+          passwordaccess = if config.genPasswd then "generate" else "prompt";
+          passworddir = ''"${config.passwdDir}"'';
+        } // lib.attrsets.optionalAttrs (config.includeExclude!="") {
+          inclexcl = ''"${pkgs.writeText "inclexcl.dsm.sys" config.includeExclude}"'';
+        }
+      );
```

Comment:
> This only works if options' effects are independent of the order in which they appear. Is that true for this config file?

I don't know -- see below for a resolution.

> It is normally easier (on future maintenance), less error prone and less likely to cause user frustration to not abstract the config file format at all and leave configuration to the user than to abstract only parts of the config file language.

I wouldn't consider that a problem here as it concerns two different configuration files with different syntax. Abstracting the "inclexcl" file would certainly be quite hard.

However, I agree with the objection above: the order of options in `dsm.sys` *is* a problem. I haven't found a statement about ordering options in `dsm.sys` in the doucmentation. Coarsely looking for order-dependent options didn't give me any suspects, but there are too many options to be sure.

Completely "unabstracting" the `dsm.sys` seems to be hard as well as we still need to place *some* options there for the system backup. I was tempted to change the type of the "extraConfig" option to contain "lines" of text instead of an "attrsOf" strings. That way users could just juse `mkBefore` and `mkAfter` to add lines in arbitrary order around those generated by the predefined config keys. That approach, however, precludes the user from overwriting or suppressing config lines generated by the predefined config keys. In light of this, I added *another* config option "text" that contains the textual result of "extraConfig" that forms the stanza. So, if the user wants to overwrite/suppress a config line that is generated by a predefined option, s/he can use "extraConfig"; if the user wants to add lines in specific order, s/he can use "text".

This solution might appear a bit over-engineered, but it's the best one I could come up with: It provides predefined options for common configuration keys and still gives the user full control over the content of each server stanza.

Reply:
You can still use `lines` to achieve your goal. A user can completely overwrite generated content by setting the variable with `mkForce`.